lib/PersonalityProvider.jsm

```diff
@@ -38,6 +74,76 @@ this.PersonalityProvider = class PersonalityProvider {
     this.scores = scores || {};
     this.interestConfig = this.scores.interestConfig;
     this.interestVector = this.scores.interestVector;
+    this.setupSyncAttachment(RECIPE_NAME);
+    this.setupSyncAttachment(MODELS_NAME);
+  }
+
+  setupSyncAttachment(collection) {
+    RemoteSettings(collection).on("sync", async event => {
+      const {
+        data: {created, updated, deleted},
+      } = event;
+
+      // Remove every removed attachment.
+      const toRemove = deleted.concat(updated.map(u => u.old));
+      await Promise.all(toRemove.map(record => this.deleteAttachment(record)));
+
+      // Download every new/updated attachment.
+      const toDownload = created.concat(updated.map(u => u.new));
+      await Promise.all(toDownload.map(record => this.downloadAttachment(record)));
+    });
+  }
+
+  async downloadAttachment(record) {
+    const {attachment: {location, filename}} = record;
+    const headers = new Headers();
+    headers.set("Accept-Encoding", "gzip");
+    const filepath = (await baseAttachmentsURL) + location;
+    const resp = await fetch(filepath, {headers});
+    if (!resp.ok) {
+      Cu.reportError(`Failed to fetch ${filepath}: ${resp.status}`);
+      return false;
+    }
+    const buffer = await resp.arrayBuffer();
+    const bytes = new Uint8Array(buffer);
+    await OS.File.makeDir(OS.Path.join(OS.Constants.Path.localProfileDir, PERSONALITY_PROVIDER_DIR_NAME));
+    const path = OS.Path.join(OS.Constants.Path.localProfileDir, PERSONALITY_PROVIDER_DIR_NAME, filename);
+    return OS.File.writeAtomic(path, bytes, {tmpPath: `${path}.tmp`});
+  }
+
+  async deleteAttachment(record) {
+    const {attachment: {filename}} = record;
+    await OS.File.makeDir(OS.Path.join(OS.Constants.Path.localProfileDir, PERSONALITY_PROVIDER_DIR_NAME));
+    const path = OS.Path.join(OS.Constants.Path.localProfileDir, PERSONALITY_PROVIDER_DIR_NAME, filename);
+
+    await OS.File.remove(path, {ignoreAbsent: true});
+    return OS.File.removeEmptyDir(OS.Path.join(OS.Constants.Path.localProfileDir, PERSONALITY_PROVIDER_DIR_NAME), {ignoreAbsent: true});
+  }
+
+  async getAttachment(record) {
+    const {attachment: {filename, hash, size}} = record;
+    await OS.File.makeDir(OS.Path.join(OS.Constants.Path.localProfileDir, PERSONALITY_PROVIDER_DIR_NAME));
+    const filepath = OS.Path.join(OS.Constants.Path.localProfileDir, PERSONALITY_PROVIDER_DIR_NAME, filename);
+    let jsonStr = "{}";
+
+    try {
+      if (!await OS.File.exists(filepath) ||
+          await OS.File.stat(filepath).size !== size ||
+          getHash(await this._getFileStr(filepath)) !== hash) {
+        await this.downloadAttachment(record);
```

Comment:
Shouldn't this be in a `while (retry < 3) {}` ?

Reply:
Agreed, do I need a similar or same retry on the sync download too?

Reply:
I'm considering moving the retry and hash check logic into downloadAttachment.

Reply:
I moved some stuff around to get the retry stuff to fit in better.